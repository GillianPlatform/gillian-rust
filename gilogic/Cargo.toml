[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }