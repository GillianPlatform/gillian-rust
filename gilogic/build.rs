fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
