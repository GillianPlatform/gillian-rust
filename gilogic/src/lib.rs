#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
