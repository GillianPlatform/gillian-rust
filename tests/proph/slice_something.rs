//@check-pass
extern crate gilogic;
use gilogic::{macros::specification, prophecies::Ownable};
use std::slice::from_raw_parts_mut;

fn test(x: &mut [u32]) -> u32 {
    x[0]
}

#[specification(
  forall m.
  requires { (x -> m) * m.own(m) }
  ensures { ret.own(m) }
)]
pub fn test2(x: *mut u32) -> u32 {
    let x = unsafe { from_raw_parts_mut(x, 1) };
    test(x)
}
