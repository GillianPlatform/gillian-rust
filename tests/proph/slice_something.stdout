#begin_init_data
{}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_pcy.gil", "i__lang_pcy.gil";

pred "<u32 as gilogic::prophecies::Ownable>::own"(+self, model):
  (0i i<=# #self) * (#self i<=# 4294967295i) * (#self == #model)
  * types(#self: Int)
  * types(#model: Int)
  * (model == #model)
  * (self == #self);


proc test(pLft_0, x) {
        m_1x := x;
        g_0 := m_1x;
        m_1x := [alloc]({{ "ref", true, {{ "slice", "u32" }} }});
        u := [store_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, g_0);
  bb0:  m_2 := 0i;
        g_2 := [load_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, true);
        g_1 := l-nth(g_2, 0i);
        m_3 := l-nth(l-nth(g_1, 0i), 1i);
        m_4 := ((m_2) i< (m_3));
        g_3 := m_4;
        u := "i__lang_assert"(g_3, "bounds check failed");
        goto bb1;
  bb1:  g_5 := [load_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, true);
        g_4 := l-nth(g_5, 0i);
        g_7 := [load_value](l-nth(l-nth(l-nth(g_4, 0i), 0i), 0i), l+(l-nth(l-nth(l-nth(g_4, 0i), 0i), 1i), {{ {{ "+", false, m_2, "u32" }} }}), "u32", true);
        g_6 := l-nth(g_7, 0i);
        m_0 := g_6;
        g_8 := m_0;
        ret := g_8;
        goto rlab;
  rlab: u := [free](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "ref", true, {{ "slice", "u32" }} }});
        return
};

spec test2(x)
[[ <value>(l-nth(#x, 0i), l-nth(#x, 1i), "u32"; #m)
  * "<u32 as gilogic::prophecies::Ownable>::own"(#m, #m)
  * types(#m: Int)
  * types(#lvar_0: Obj, #lvar_1: List) * (#x == {{ #lvar_0, #lvar_1 }})
  * (x == #x) ]]
[[ "<u32 as gilogic::prophecies::Ownable>::own"(ret, #m) ]]
normal

proc test2(x) {
        m_1x := x;
        m_2x := [alloc]({{ "ref", true, {{ "slice", "u32" }} }});
        m_3 := [alloc]({{ "ref", true, {{ "slice", "u32" }} }});
  bb0:  m_4 := m_1x;
        g_0 := [new_lft]();
        pLft_6 := l-nth(g_0, 0i);
        g_1 := m_4;
        g_2 := "std::slice::from_raw_parts_mut"(pLft_6, "u32", g_1, 1i);
        u := [store_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, g_2);
        goto bb1;
  bb1:  g_4 := [load_value](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, true);
        g_3 := l-nth(g_4, 0i);
        g_5 := [new_lft]();
        pLft_5 := l-nth(g_5, 0i);
        u := [store_value](l-nth(m_2x, 0i), l-nth(m_2x, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, g_3);
        g_7 := [load_value](l-nth(m_2x, 0i), l-nth(m_2x, 1i), {{ "ref", true, {{ "slice", "u32" }} }}, true);
        g_6 := l-nth(g_7, 0i);
        g_8 := [new_lft]();
        pLft_7 := l-nth(g_8, 0i);
        m_5 := g_6;
        g_9 := m_5;
        g_10 := "test"(pLft_6, g_9);
        m_0 := g_10;
        goto bb2;
  bb2:  g_11 := m_0;
        ret := g_11;
        goto rlab;
  rlab: u := [free](l-nth(m_2x, 0i), l-nth(m_2x, 1i), {{ "ref", true, {{ "slice", "u32" }} }});
        u := [free](l-nth(m_3, 0i), l-nth(m_3, 1i), {{ "ref", true, {{ "slice", "u32" }} }});
        return
};

